server_environment/models/server_env_tech_name_mixin.py

```diff
@@ -25,41 +26,36 @@ class ServerEnvTechNameMixin(models.AbstractModel):
     _sql_constraints = [
         ("tech_name_uniq", "unique(tech_name)", "`tech_name` must be unique!",)
     ]
-    # TODO: could leverage the new option for computable / writable fields
-    # and get rid of some onchange / read / write code.
+
     tech_name = fields.Char(
+        string="Technical Name",
         help="Unique name for technical purposes. Eg: server env keys.",
+        compute="_compute_tech_name",
+        store=True,
+        readonly=False,
     )
 
     _server_env_section_name_field = "tech_name"
 
-    @api.onchange("name")
-    def _onchange_name_for_tech(self):
-        # Keep this specific name for the method to avoid possible overrides
-        # of existing `_onchange_name` methods
-        if self.name and not self.tech_name:
-            self.tech_name = self.name
-
-    @api.onchange("tech_name")
-    def _onchange_tech_name(self):
-        if self.tech_name:
-            # make sure is normalized
-            self.tech_name = self._normalize_tech_name(self.tech_name)
-
-    @api.model
-    def create(self, vals):
-        self._handle_tech_name(vals)
-        return super().create(vals)
-
-    def write(self, vals):
-        self._handle_tech_name(vals)
-        return super().write(vals)
-
-    def _handle_tech_name(self, vals):
-        # make sure technical names are always there
-        if not vals.get("tech_name") and vals.get("name"):
-            vals["tech_name"] = self._normalize_tech_name(vals["name"])
+    @api.depends("name")
+    def _compute_tech_name(self):
+        for rec in self:
+            # Update tech_name only if it hasn't been set or if we're
+            # dealing with a new record.
+            if not rec.tech_name or not rec._origin.id:
+                rec.tech_name = self._normalize_tech_name(rec.name)
+
+    @api.constrains("tech_name")
+    def _check_tech_name(self):
+        for rec in self.filtered("tech_name"):
```

Comment:
I'd replace this w/ an inverse method that normalizes ALWAYS.

Reply:
What I don't like about normalizing automatically on an inverse method is that if you write to the record programatically you'll get a silent automatic formatting, which is misleading.

ie:

```py
rec = self.env["your.model"].create({"name": "Test 123", "tech_name": "test 123"})
rec.tech_name # test_123
```

For the same reason I like to keep the constraint

What about restoring the onchange instead?

If the record is indeed exposed to the form view and written on directly, the onchange can format it automatically. We can even display a warning message.

Check my latest commit

Reply:
> What I don't like about normalizing automatically on an inverse method is that if you write to the record programatically you'll get a silent automatic formatting, which is misleading.

I know, but in this case I don't want to trust the user at all. At least, that's my POV ;)
 
> For the same reason I like to keep the constraint
> 
> What about restoring the onchange instead?
> 
> If the record is indeed exposed to the form view and written on directly, the onchange can format it automatically. We can even display a warning message.
> 
> Check my latest commit

I don't like the warning. As mentioned above: we should not trust the user and the user should spend time in understanding how we want that field. This field is technical and we want take full control on it.

IMO What is relevant for the user is that the final value is displayed. That's all the user needs to know.

With this in mind, the onchange is useless if you normalize it automatically.

My $0.02

Reply:
you can keep the constraint if you prefer ;)

Reply:
> I don't like the warning. As mentioned above: we should not trust the user and the user should not spend time in understanding how we want that field. This field is technical and we want take full control on it.

It's not just a warning, the onchange is also taking care of normalizing the value
The warning is to inform the user that the value has changed, rather than doing it silently (simply to play nice and not to be misleading)

As the onchange doesn't play when writing on records programatically, this enables the constraint.
Otherwise if we use inverse, the constraint will never really play and the result can be misleading in the programatic scenario.

I think how it is right now covers both scenarios: automatic on forms, exception on orm
:white_check_mark: In both cases the user isn't trusted (constraint)
:white_check_mark: In forms the field is automatically normalized (user doesn't spend time in understanding how we want the field)
:white_check_mark:  Cherry on top: warning adverting the field has been normalized 

As for how it should be exposed in forms (which is optional in fact, it could very well be hidden), I think something like this is appropiate (only **Administration / System** in debug mode):
https://github.com/OCA/server-env/blob/1c1a0b770309b3e77ebf7665a88bb92522e817c9/payment_environment/views/payment_acquirer.xml#L8-L17

Afterall this field is only meant to be seen / changed by sys admins.